initial pass at sorted-and-binned render phase items (#1491)
bevyengine/bevy-website: release-content/0.14/migration-guides/12453_13277_14029_sorted_and_binned_render_phase_items.md

> Usage of `PhaseItem` has been split into `BinnedPhaseItem` and `SortedPhaseItem`. If you have custom `PhaseItem`s you will need to choose one of the new types. Notably some phases _must_ be Sorted (such as Transparent and Transmissive), while others can be Binned. Effectively Sorted is "what Bevy did before" and Binned is new, and the point of this change is to avoid sorting when possible for improved performance.

Reviewer: ```suggestion
Usage of `PhaseItem` has been split into `BinnedPhaseItem` and `SortedPhaseItem`. If you have custom `PhaseItem`s, you will need to choose one of the new types. Notably, some phases _must_ be sorted (such as transparent and transmissive), while others can be binned. Effectively, sorted is "what Bevy did before" while binned is a new method that places batchable items together, improving performance. Note that binned phase items largely remove z-fighting.
```

Just a few wording improvements that explains what binned phase items do, and de-capitalizes some words. (Did you have a specific reason for doing so?)

Reviewer 2: I don't understand why binned phase item would remove z-fighting. The sentence seems to imply that z-fighting exists with sorted and is fixed with binned, which is quite unintuitive.

Reviewer: I mentioned it because of something @mockersf wrote, though I have not experienced it myself. Perhaps it should be removed.

Author: reasoning behind capitalization was that Transparent has `Transparent2d`, `Transparent3d`, and `TransparentUi` and I was trying to convey that this is "A Real Thing" and not "a vague idea" without writing them all out. Not terribly important though.

Reviewer 2: Maybe use "(such as  `Transparent2d`, `Transmissive3d`, _etc._)"?

> Usage of `PhaseItem` has been split into `BinnedPhaseItem` and `SortedPhaseItem`. If you have custom `PhaseItem`s you will need to choose one of the new types. Notably some phases _must_ be Sorted (such as Transparent and Transmissive), while others can be Binned. Effectively Sorted is "what Bevy did before" and Binned is new, and the point of this change is to avoid sorting when possible for improved performance.

Reviewer 3: ```suggestion
Usage of `PhaseItem` has been split into `BinnedPhaseItem` and `SortedPhaseItem`. If you have custom `PhaseItem`s you will need to choose one of these types. Notably some phases _must_ be Sorted (such as Transparent and Transmissive), while others can be Binned. Effectively Sorted is "what Bevy did before" and Binned is new, and the point of this change is to avoid sorting when possible for improved performance.
```

> If you're looking for a quick migration, consider picking [`SortedPhaseItem`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/trait.SortedPhaseItem.html) which requires the fewest code changes.

Reviewer: ```suggestion
If you're looking for a quick migration, consider picking [`SortedPhaseItem`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/trait.SortedPhaseItem.html) which requires the fewest code changes.
```

At this point, you can link directly to the 0.14 docs.

> If you're looking for higher performance (and your phase doesn’t require sorting) you may want to pick [`BinnedPhaseItem`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/trait.BinnedPhaseItem.html). Notably bins are populated based on `BinKey` and everything in the same bin is potentially batchable.

Reviewer: ```suggestion
If you're looking for higher performance (and your phase doesn’t require sorting) you may want to pick [`BinnedPhaseItem`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/trait.BinnedPhaseItem.html). Notably bins are populated based on `BinKey` and everything in the same bin is potentially batchable.
```

> If you're looking for higher performance (and your phase doesn’t require sorting) you may want to pick [`BinnedPhaseItem`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/trait.BinnedPhaseItem.html). Notably bins are populated based on `BinKey` and everything in the same bin is potentially batchable.

Reviewer 2: "is _potentially_ batchable" or is actually batched (binned)?

Author: "potentially batchable" is the wording used in the original PR, so I deferred to that - https://github.com/bevyengine/bevy/pull/12453

If you are only consuming these types, then a `Query` for a type like `&mut RenderPhase<Transparent2d>` will become a `Resource` as such:

```rust
mut transparent_render_phases: ResMut<ViewSortedRenderPhases<Transparent2d>>
```

`ViewSortedRenderPhases` and `ViewBinnedRenderPhases` are used in accordance with which phase items you're trying to access (sorted or binned).

> Examples of [`SortedPhaseItems`s](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/trait.SortedPhaseItem.html#implementors):

Reviewer: ```suggestion
Examples of [`SortedPhaseItems`s](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/trait.SortedPhaseItem.html#implementors):
```

- Transmissive3d
- Transparent2d
- Transparent3d
- TransparentUi

> Examples of [`BinnedPhaseItem`s](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/trait.BinnedPhaseItem.html#implementors) include:

Reviewer: ```suggestion
Examples of [`BinnedPhaseItem`s](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/trait.BinnedPhaseItem.html#implementors) include:
```

> - Opaque3d

Reviewer 2: Link them all maybe? (sorry for the extra work 😅)

Author: Could do. I linked Shadow specifically because to someone less familiar it looks out of place next to things that are clearly 2d/3d forward/deferred/prepass related. The full list is already available at the "Examples of" links.

- Opaque3dPrepass
- Opaque3dDeferred
- AlphaMask3d
- AlphaMask3dPrepass
- AlphaMask3dDeferred

> - [Shadow](https://docs.rs/bevy/0.14.0-rc.4/bevy/pbr/struct.Shadow.html)

Reviewer: ```suggestion
- [Shadow](https://docs.rs/bevy/0.14.0/bevy/pbr/struct.Shadow.html)
```

> If you do not have a mesh (such as for GPU-driven particles or procedural generation) and want to use the new binned behavior, the [`BinnedRenderPhase`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html) includes a new `non_mesh_items` collection which correlates with a new [`BinnedRenderPhaseType`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html). This type is used when [add](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html#method.add)ing items to the `BinnedRenderPhase`.

Reviewer 3: ```suggestion
If you do not have a mesh (such as for GPU-driven particles or procedural generation) and want to use the new binned behavior, the [`BinnedRenderPhase`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html) includes a `non_mesh_items` collection which correlates with the [`BinnedRenderPhaseType`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html). This type is used when [add](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html#method.add)ing items to the `BinnedRenderPhase`.
```

> If you do not have a mesh (such as for GPU-driven particles or procedural generation) and want to use the new binned behavior, the [`BinnedRenderPhase`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html) includes a new `non_mesh_items` collection which correlates with a new [`BinnedRenderPhaseType`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html). This type is used when [add](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html#method.add)ing items to the `BinnedRenderPhase`.

Reviewer: ```suggestion
If you do not have a mesh (such as for GPU-driven particles or procedural generation) and want to use the new binned behavior, the [`BinnedRenderPhase`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html) includes a new `non_mesh_items` collection which correlates with a new [`BinnedRenderPhaseType`](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html). This type is used when [add](https://docs.rs/bevy/0.14.0/bevy/render/render_phase/struct.BinnedRenderPhase.html#method.add)ing items to the `BinnedRenderPhase`.
```

> If you do not have a mesh (such as for GPU-driven particles or procedural generation) and want to use the new binned behavior, the [`BinnedRenderPhase`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html) includes a new `non_mesh_items` collection which correlates with a new [`BinnedRenderPhaseType`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html). This type is used when [add](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html#method.add)ing items to the `BinnedRenderPhase`.

Reviewer 2: These are not new, the type and enum didn't exist before.

```suggestion
If you do not have a mesh (such as for GPU-driven particles or procedural generation) and want to use the new binned behavior, the [`BinnedRenderPhase`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html) includes a `non_mesh_items` collection which correlates with a [`BinnedRenderPhaseType`](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html). This type is used when [add](https://docs.rs/bevy/0.14.0-rc.4/bevy/render/render_phase/struct.BinnedRenderPhase.html#method.add)ing items to the `BinnedRenderPhase`.
```
